### PR TITLE
=act Short path if message clazz is classOf[java.lang.Object]

### PR DESCRIPTION
**Pull request by He-Pin**

### CODE REVIEW

**Review by raboof — COMMENTED** [review]
Do we have any benchmarks that exercise this code? This seems like an attractive optimization, but it'd be good to double-check its effect ;). Otherwise LGTM

**Comment by hepin1989** [comment]
I can submit a PR with a benchmark when I found any time, but current, I must do the recruiting thing.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala [hunk 1]
```diff
@@ -76,10 +76,11 @@ private[akka] final class InterceptorImpl[O, I](
     new InterceptorImpl(interceptor, newNested)
 
   override def receive(ctx: typed.TypedActorContext[O], msg: O): Behavior[O] = {
-    // TODO performance optimization could maybe to avoid isAssignableFrom if interceptMessageClass is Class[Object]?
     val interceptMessageClass = interceptor.interceptMessageClass
     val result =
-      if ((interceptMessageClass ne null) && interceptor.interceptMessageClass.isAssignableFrom(msg.getClass))
+      if ((interceptMessageClass ne null) &&
+          (interceptMessageClass == classOf[java.lang.Object] ||
```
Comment (hepin1989) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:82]: add this

**Comment by johanandren** [comment]
Thanks @hepin1989, let's hold off this change until we have benchmark proof it helps.

**Comment by hepin1989** [comment]
Yes, Let's hold this up. I will update with it this weekend.

**Comment by hepin1989** [comment]
run with 
```
jmh:run -i 10 -wi 10 -f1 -t1 .*akka.actor.typed.internal.InterceptorImplBench.*
```

```
[info] Benchmark                                     Mode  Cnt          Score         Error  Units
[info] InterceptorImplBench.benchReceiveOpt         thrpt   10  690834550.948 ± 8863426.234  ops/s
[info] InterceptorImplBench.benchReceiveWithoutOpt  thrpt   10  404297496.845 ± 3617637.606  ops/s
```

**Comment by johanandren** [comment]
Hmm, very *micro* micro bench there, was kind of hoping on something closer to actual usage in actors. I wonder if it makes any difference in practice. 

Looking a bit at where it would have effect in practice, we have `Any`/`Object` as the `Outer` type in the supervisor impl (where it may help), the logger interceptor (where logging will likely be more expensive), or when the user intercepts a `Behavior[Any]` which isn't really _that_ likely.

I think the supervision could be good enough as a motivator to have this optimization in.

@hepin1989 We'd want the benchmark to at least test the actual implementation rather than two variations inlined in the test, do you think you can update that?

**Comment by hepin1989** [comment]
Will if I find some time.

**Comment by johanandren** [comment]
I rewrote the benchmark to one that runs an actor with the interceptor in https://github.com/akka/akka/commit/e70929c57d5fcbec4ad566454b20735bd54d066c

Here are the numbers from actual use,

with optimization
```
[info] Benchmark                         (whichBench)   Mode  Cnt        Score        Error  Units
[info] InterceptorBenchmark.bench            baseline  thrpt   10  3484511,183 ± 386990,479  ops/s
[info] InterceptorBenchmark.bench      anyPassthrough  thrpt   10  3140576,552 ± 273819,385  ops/s
[info] InterceptorBenchmark.bench  subtypePassthrough  thrpt   10  3045528,708 ± 174725,171  ops/s
[info] InterceptorBenchmark.bench          supervised  thrpt   10  3089868,034 ± 163507,129  ops/s
```

master, no optimization
```
[info] Benchmark                         (whichBench)   Mode  Cnt        Score        Error  Units
[info] InterceptorBenchmark.bench            baseline  thrpt   10  3595361,551 ± 426977,183  ops/s
[info] InterceptorBenchmark.bench      anyPassthrough  thrpt   10  3098631,911 ± 224138,146  ops/s
[info] InterceptorBenchmark.bench  subtypePassthrough  thrpt   10  3179245,809 ± 159478,233  ops/s
[info] InterceptorBenchmark.bench          supervised  thrpt   10  3068362,370 ± 212736,924  ops/s
```

The improvement looks like it is inside of the margin of error, so I am not entirely convinced it is worth merging this. It's a bit noisy though, for example the subtype that does trigger a passthrough is faster than object for some reason. Could perhaps be that assignableFrom for the exact type is pretty fast.

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala (hunk 1 above)
Comment (hepin1989) [reply akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:83]: refs https://stackoverflow.com/revisions/26514984/10

**Comment by hepin1989** [comment]
https://stackoverflow.com/questions/496928/what-is-the-difference-between-instanceof-and-class-isassignablefrom?noredirect=1&lq=1

**Comment by johanandren** [comment]
I know the difference in how costly isAssignableFrom vs a simple equals on class, what I am saying is that it looks like that cost is potentially not important in the actual use case (40 000 msg/s improvement when the error is >200 000 msg/s)

My gut feeling was that it would make a difference, but it doesn't look like that.

**Comment by johanandren** [comment]
Let's close this for now and re-open if some evidence this change would matter in the grander scheme of things surfaces.

**Pull request closed without merging**